Port OgComplexWidgetTest

pfrenssen opened the pull request:
Some bugs are popping up when adding groups to the group audience field using the complex widget. One of these that was recently fixed is #138.

Let's port the original OgComplexWidgetTest so that we can discover any remaining problems, and ensure it keeps working in the future.

jhedstrom committed:
Split OgComplexWidgetTest into a separate class.

pfrenssen committed:
Remove the original D7 test.

pfrenssen committed:
Put the test in a new folder dedicated to web tests.

pfrenssen committed:
Use shorthand array syntax.

pfrenssen committed:
Start porting.

pfrenssen commented:
Started porting. I'm currently stuck on the audience field not being shown in the node edit form by default. I'm trying to make it visible but I'm getting schema errors when I do so:

```
Drupal\Core\Config\Schema\SchemaIncompleteException: Schema errors for core.entity_form_display.node.post.default with the following errors:
core.entity_form_display.node.post.default:content.og_group_ref.settings.match_operator missing schema
core.entity_form_display.node.post.default:content.og_group_ref.settings.size missing schema
core.entity_form_display.node.post.default:content.og_group_ref.settings.placeholder missing schema
in Drupal\Core\Config\Testing\ConfigSchemaChecker->onConfigSave() (line 98 of core/lib/Drupal/Core/Config/Testing/ConfigSchemaChecker.php).
```

pfrenssen committed:
Add missing schema for the complex widget.

amitaibu commented:
> I'm currently stuck on the audience field not being shown in the node edit form by default

Yes, we should take care of it in `Og::createField()`

Thanks for working on it :)

pfrenssen commented:
OK so the test is correct, the schema is simply missing.

pfrenssen commented:
@amitaibu good to know, I was wondering about this. When I continue tomorrow I will create a ticket for that and refer to it in the code.

pfrenssen commented:
The test is not running, it should be added to Travis.

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into port-og-complex-widget-test

pfrenssen committed:
Enable functional tests on Travis.

These changes are inspired by the excellent Travis integration of the Rules module.

pfrenssen commented:
Oops I sent Travis into an endless loop :-/

pfrenssen committed:
Attempt to fix endless loop on Travis CI.

pfrenssen committed:
The group audience field will be visible in the future.

pfrenssen committed:
Let's make it a BrowserTest.

pfrenssen committed:
Use various traits to import test helper methods.

pfrenssen committed:
Porting in progress.

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into port-og-complex-widget-test

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into port-og-complex-widget-test

pfrenssen committed:
Merge branch 'target-all-bundles' into port-og-complex-widget-test

pfrenssen committed:
Porting in progress.

pfrenssen commented:
I merged in #147 since that is a bug that was discovered while working on this and I need the fix to continue here.

pfrenssen committed:
The 'Other Groups' field was looking at the wrong key and letting empty values through. Started on a test for this field.

pfrenssen commented:
I've been doing manual testing and noticed some problems with the 'Other groups' field:
- If the 'Other Groups' field has a value but 'Groups Audience' doesn't, then an entity with ID `NULL` is being used for the Groups Audience field, throwing a fatal error.
- The form key for the 'Other Groups' field was wrong, it was looking in `$form['other_groups']` instead of in `$form['og_group_ref']['other_groups']`.
- For some reason the standard `ValidReferenceConstraintValidator` is running again rather than `ValidOgMembershipReferenceConstraintValidator`, resulting in this error:
  
  > TypeError: Argument 1 passed to Drupal\Core\Field\Plugin\Field\FieldWidget\EntityReferenceAutocompleteWidget::errorElement() must be of the type array, null given, called in core/lib/Drupal/Core/Field/WidgetBase.php on line 444 in Drupal\Core\Field\Plugin\Field\FieldWidget\EntityReferenceAutocompleteWidget->errorElement() (line 122 of core/lib/Drupal/Core/Field/Plugin/Field/FieldWidget/EntityReferenceAutocompleteWidget.php).

Committed work in progress, I have to work on something else for the time being.

pfrenssen committed:
Add reminder for my future self.

pfrenssen committed:
Finish OgComplexWidgetTest::testOtherGroups().

pfrenssen committed:
Merge branch 'other-groups-no-references' into port-og-complex-widget-test

pfrenssen committed:
Porting in progress.

pfrenssen committed:
Small code style fixes in Og.

pfrenssen committed:
Finish porting of ::testHiddenGroupIds(). This functionality has regressed.

pfrenssen commented:
Completed the `testHiddenGroupIds()` test. It tests that if a user edits a node that references a group the user does not have access to, the group reference is not lost when the user submits the form.

This currently fails, so this functionality has apparently regressed.

amitaibu commented:
> This currently fails, so this functionality has apparently regressed.

Indeed we didn't tackle it yet. I think it's fine not keep it out for now, so this PR doesn't become too scope intensive.

pfrenssen commented:
@amitaibu I intend to fix the bugs that I find and split them off to separate tickets so they can be reviewed separately if that is OK for you? That will reduce the scope of this over time, so in the end it will be trivial to review.

amitaibu commented:
:+1:

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into port-og-complex-widget-test

pfrenssen committed:
Merge branch 'reference-string-ids' into port-og-complex-widget-test

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into port-og-complex-widget-test-wip

pfrenssen committed:
Fix MissingGroupException when loading ReferenceStringIdTest.

pfrenssen committed:
Use a random field name for the group audience field.

Easier to type, and consistent with the other tests.

pfrenssen committed:
Content is now referenced solely through the ER field. OgMembership is for users only.

pfrenssen committed:
Use the default field name again, the complex field was disappearing.

amitaibu commented:
Is this ready for review?

pfrenssen commented:
No not yet, there is still a lot to do here.

pfrenssen commented:
This is currently postponed on the big refactoring work that is happening on the OgComplex widget in #161. Once that is done we can work on the `testFieldModes()` test, and progress from there.
